libbpf-tools/memleak.bpf.c

```diff
@@ -0,0 +1,237 @@
+	__type(key, __u64);
+	__type(value, struct alloc_info_t);
+} allocs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u64);
+	__type(value, __u64);
+} memptrs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(key_size, sizeof(__u32));
+} stack_traces SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, __u64);
+	__type(value, struct combined_alloc_info_t);
+} combined_allocs SEC(".maps");
+
+static void incr_stat(__u64 stack_id, __u64 sz) {
+	struct combined_alloc_info_t *existing_cinfo;
+	struct combined_alloc_info_t cinfo = {0};
+
+	existing_cinfo = bpf_map_lookup_elem(&combined_allocs, &stack_id);
+	if (existing_cinfo) {
+		cinfo = *existing_cinfo;
+	}
+
+	cinfo.total_size += sz;
+	cinfo.number_of_allocs += 1;
+	bpf_map_update_elem(&combined_allocs, &stack_id, &cinfo, BPF_ANY);
+}
+
+static void decr_stat(__u64 stack_id, __u64 sz) {
+	struct combined_alloc_info_t *existing_cinfo;
+	struct combined_alloc_info_t cinfo = {0};
+
+	existing_cinfo = bpf_map_lookup_elem(&combined_allocs, &stack_id);
+	if (existing_cinfo) {
+		cinfo = *existing_cinfo;
+	}
+
+	if (sz >= cinfo.total_size) {
+		cinfo.total_size = 0;
+	} else {
+		cinfo.total_size -= sz;
+	}
+	if (cinfo.number_of_allocs > 0) {
+		cinfo.number_of_allocs -= 1;
+	}
+	bpf_map_update_elem(&combined_allocs, &stack_id, &cinfo, BPF_ANY);
+}
+
+static int handle_free_entry(void *ptr) {
+	__u64 addr = (__u64)ptr;
+	struct alloc_info_t *info = bpf_map_lookup_elem(&allocs, &addr);
+
+	if (!info) {
+		return 0;
+	}
+
+	if (debug_enabled) {
+		bpf_trace_printk("free entered, address = %lx, size = %lu\n", addr, info->size);
+	}
+	bpf_map_delete_elem(&allocs, &addr);
+	decr_stat(info->stack_id, info->size);
+	return 0;
+}
+
+static int handle_alloc_entry(size_t size) {
+	if (filter_min_size && size < min_size) {
+		return 0;
+	}
+	if (filter_max_size && size > max_size) {
+		return 0;
+	}
+	if (sample_rate > 1) {
+		__u64 ts = bpf_ktime_get_ns();
+		if (ts % sample_rate != 0)
+			return 0;
+	}
+
+	__u64 pid = bpf_get_current_pid_tgid();
+	__u64 size64 = size;
+	bpf_map_update_elem(&sizes, &pid, &size64, BPF_ANY);
+
+	if (debug_enabled) {
+		bpf_trace_printk("alloc entered, size = %u\n", size);
+	}
+	return 0;
+}
+
+static int handle_alloc_exit(void *ctx, __u64 addr) {
+	__u64 pid = bpf_get_current_pid_tgid();
+	__u64 *size64 = bpf_map_lookup_elem(&sizes, &pid);
+	struct alloc_info_t info = {0};
+
+	if (size64 == 0) {
+		return 0; // missed alloc entry
+	}
+
+	info.size = *size64;
+	bpf_map_delete_elem(&sizes, &pid);
+
+	if (addr) {
+		info.timestamp_ns = bpf_ktime_get_ns();
+		info.stack_id = bpf_get_stackid(ctx, &stack_traces, stack_flags);
+		bpf_map_update_elem(&allocs, &addr, &info, BPF_ANY);
+		incr_stat(info.stack_id, info.size);
+	}
+
+	if (debug_enabled) {
+		bpf_trace_printk("alloc exited, size = %lu, result = %lx\n", info.size, addr);
```

Comment:
this is the problem, you can't just use strings directly with bpf_trace_printk(). Either use bpf_printk(), which is a macro that will put such string into on stack array, or you can make this string a static variable:

```
static const char[] fmt = "alloc exited, size = %lu, result = %lx\n";
```

Reply:
Ah, thanks.